config/jupyterhub_config.py

```diff
@@ -4,24 +4,40 @@
 https://jupyterhub.readthedocs.io/en/latest/tutorial/getting-started/config-basics.html
 https://jupyterhub.readthedocs.io/en/stable/reference/config-reference.html
 """
+
 import os
 
+import nativeauthenticator
+
 from jupyterhub_config.custom_spawner import VirtualEnvSpawner
 
 c = get_config()
 
-# Set the authenticator class
-# TODO: Change the authenticator class to a secure one (e.g. GitHubOAuthenticator)
-c.JupyterHub.authenticator_class = 'jupyterhub.auth.DummyAuthenticator'
-c.Authenticator.allowed_users = {'spark_user', 'test_user1', 'test_user2'}
-c.DummyAuthenticator.password = os.environ['JUPYTERHUB_ADMIN_PASSWORD']
+# Set the authenticator class to nativeauthenticator
+# ref: https://native-authenticator.readthedocs.io/en/latest/quickstart.html
+c.JupyterHub.authenticator_class = 'native'
+c.JupyterHub.template_paths = [f"{os.path.dirname(nativeauthenticator.__file__)}/templates/"]
+c.NativeAuthenticator.open_signup = True
+c.NativeAuthenticator.check_common_password = True
+c.NativeAuthenticator.minimum_password_length = 6
 
+# Set up the admin user
 c.Authenticator.admin_users = {'spark_user'}
+# TODO set admin user password to os.environ['JUPYTERHUB_ADMIN_PASSWORD'] automatically - currently spark_user is created manually with the signup page
```

Comment:
Is there anything preventing a user from creating their name as spark_user other than an admin doing it first via the uI?

Reply:
they cannot. The auth class checks current db to make sure there is no existing user.

Reply:
they don't. We create users if they don't exist at startup.

Reply:
Ok, but the admin users are created before any users have a chance to create an account?

Reply:
no. It just happened we already created `spark_user` in Dockerfile. If the admin user doesn't exist, it will go through user creation code. The username must be the same as what's listed in `c.Authenticator.admin_users` to be an admin.

Reply:
So theoretically, if user X came along fast after the container came up, they could create an account called `spark_user` an be an admin?

Reply:
no, I don't think so. `spark_user` is already created by dockerfile.

Reply:
But the juypterhub auth system doesn't have an account set up for that yet. So what prevents someone other than you (or whoever) from taking it?

Reply:
taking what? I am completely lost here. What's the concern here? Someone will create a `spark_user`? Before the first time the user_db gets created?

Reply:
The concern is that the hub container starts and some user comes along and creates a `spark_user` account, and thus gets admin rights. It's not clear to me whether that can happen or not, and if not, why not

Reply:
yea. It can happen before we manually created the user. I don't see it's a concern or something likely to happen. I have a todo item to set the admin password. Also once the admin user is created, it won't happen. So theoretically,  it can happen, I don't see a way to prevent it. I can move locally created user db to the mounted dir so that auth read from existing db.

Reply:
If you want to leave it as is for now but create an issue that's fine with me

>  I can move locally created user db to the mounted dir so that auth read from existing db.

Don't you need to do this to preserve user account over container recreations?

Reply:
no. Users will signup by themselves.

Reply:
I don't understand how this answer pertains to the question

Reply:
> Don't you need to do this to preserve user accounts over container recreations?

Users will signup and jupyterhub service will create an account accordingly in `jupyterhub.sqlite` and spawner will create a system user accordingly. So I don't need to move locally created db to the mounted dir.

Reply:
But if the db isn't in a mounted dir, everytime the container is recreated all the user accounts will be wiped

Reply:
yea. db is in a mounted dir. I was saying I can copy the locally created db which has spark_user records to the mounted dir before the first time container starts. That would prevent some random user create spark_user.

Reply:
Ok, now I'm confused as to why there are 2 db locations, the mounted location and the 'locally created' location. Shouldn't there be just one location?

Reply:
no. In my local environment (on my laptop), I have created a db for testing purposes. And that db has `spark_user` records. I was saying I could copy that db to the mounted location in the server, so that before container starts, the db in the mounted location exists.

Reply:
Oh, you add a prefilled db with the admin accounts to the container, I see.

Ok. If you want to do that that sounds reasonable (unless there's some drawback like blowing away user settings that are added to the db later from the UI), but I don't think it's required since this is a temporary solution and presumably won't be doing to prod.